Source/Core/VideoBackends/Software/TextureSampler.cpp

```diff
@@ -112,64 +112,109 @@ void Sample(s32 s, s32 t, s32 lod, bool linear, u8 texmap, u8* sample)
   }
 }
 
-void SampleMip(s32 s, s32 t, s32 mip, bool linear, u8 texmap, u8* sample)
+constexpr u32 maxMips = 10;
+
+struct SamplerState
 {
-  auto texUnit = bpmem.tex.GetUnit(texmap);
+  u32 width;
+  u32 height;
+  u8* pointer;
+  u8* pointer_odd;
+  u8* pointer_tlut;
+  u32 mipOffsets[maxMips + 1];
+  TextureFormat texfmt;
+  TLUTFormat tlutfmt;
+  bool manually_managed;
+  WrapMode wrap_s;
+  WrapMode wrap_t;
+};
+
+static SamplerState Units[8];
+
+void SetupSampler(u8 texmap)
+{
+  auto& state = Units[texmap];
+  auto& texUnit = bpmem.tex.GetUnit(texmap);
 
   const TexMode0& tm0 = texUnit.texMode0;
   const TexImage0& ti0 = texUnit.texImage0;
   const TexTLUT& texTlut = texUnit.texTlut;
-  const TextureFormat texfmt = ti0.format;
-  const TLUTFormat tlutfmt = texTlut.tlut_format;
+  state.texfmt = ti0.format;
+  state.tlutfmt = texTlut.tlut_format;
 
-  const u8* imageSrc;
-  const u8* imageSrcOdd = nullptr;
```

Comment:
`state.pointer_odd` should probably be set to `nullptr` here just to be safe, as otherwise it doesn't get updated from the previous value.